lib/upgrade/etcd/conversionv1v3/workloadendpoint.go

```diff
@@ -0,0 +1,258 @@
+// limitations under the License.
+
+package conversionv1v3
+
+import (
+	"strings"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/projectcalico/libcalico-go/lib/apis/v1/unversioned"
+	"github.com/projectcalico/libcalico-go/lib/backend/model"
+	apiv1 "github.com/projectcalico/libcalico-go/lib/apis/v1"
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/net"
+	"github.com/projectcalico/libcalico-go/lib/names"
+)
+
+type WorkloadEndpoint struct{}
+
+// APIV1ToBackendV1 converts v1 WorkloadEndpoint API to v1 WorkloadEndpoint KVPair.
+func (_ WorkloadEndpoint) APIV1ToBackendV1(rIn unversioned.Resource) (*model.KVPair, error) {
+	ah := rIn.(apiv1.WorkloadEndpoint)
+	k, err := convertMetadataToKey(ah.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	// IP networks are stored in the datastore in separate IPv4 and IPv6
+	// fields.  We normalise the network to ensure the IP is correctly
+	// masked.
+	ipv4Nets := []net.IPNet{}
+	ipv6Nets := []net.IPNet{}
+	for _, n := range ah.Spec.IPNetworks {
+		n = *(n.Network())
+		if n.Version() == 4 {
+			ipv4Nets = append(ipv4Nets, n)
+		} else {
+			ipv6Nets = append(ipv6Nets, n)
+		}
+	}
+
+	ipv4NAT := []model.IPNAT{}
+	ipv6NAT := []model.IPNAT{}
+	for _, n := range ah.Spec.IPNATs {
+		nat := model.IPNAT{IntIP: n.InternalIP, ExtIP: n.ExternalIP}
+		if n.InternalIP.Version() == 4 {
+			ipv4NAT = append(ipv4NAT, nat)
+		} else {
+			ipv6NAT = append(ipv6NAT, nat)
+		}
+	}
+
+	var ports []model.EndpointPort
+	for _, port := range ah.Spec.Ports {
+		ports = append(ports, model.EndpointPort{
+			Name:     port.Name,
+			Protocol: port.Protocol,
+			Port:     port.Port,
+		})
+	}
+
+	d := model.KVPair{
+		Key: k,
+		Value: &model.WorkloadEndpoint{
+			Labels:           ah.Metadata.Labels,
+			ActiveInstanceID: ah.Metadata.ActiveInstanceID,
+			State:            "active",
+			Name:             ah.Spec.InterfaceName,
+			Mac:              ah.Spec.MAC,
+			ProfileIDs:       ah.Spec.Profiles,
+			IPv4Nets:         ipv4Nets,
+			IPv6Nets:         ipv6Nets,
+			IPv4NAT:          ipv4NAT,
+			IPv6NAT:          ipv6NAT,
+			IPv4Gateway:      ah.Spec.IPv4Gateway,
+			IPv6Gateway:      ah.Spec.IPv6Gateway,
+			Ports:            ports,
+		},
+		Revision: ah.Metadata.Revision,
+	}
+
+	return &d, nil
+}
+
+// BackendV1ToAPIV3 converts v1 WorkloadEndpoint KVPair to v3 API.
+func (_ WorkloadEndpoint) BackendV1ToAPIV3(kvp *model.KVPair) (Resource, error) {
+	wepKey, ok := kvp.Key.(model.WorkloadEndpointKey)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid WorkloadEndpoint resource key")
+	}
+	wepValue, ok := kvp.Value.(model.WorkloadEndpoint)
+	if !ok {
+		return nil, fmt.Errorf("value is not a valid WorkloadEndpoint resource Value")
+	}
+
+	labels := convertLabels(wepValue.Labels)
+	namespace := labels["projectcalico.org/namespace"]
+	pod := getPodName(wepKey.WorkloadID, namespace)
```

Comment:
Whether we have a Pod, ContainerID or WorkloadID depends on the orchestrator.  Whilst we don't need this logic now I think it should be relatively simple to implement now.

For k8s, there is no WorkloadID.
For CNI, the workloadID -> ContainerID, (workloadID is left blank)
For libnetwork, the workloadID is libnetwork(Pod + containerID are blank)
For everything else, the workloadID == workloadID in v3 (Pod + containerID are blank)

Reply:
if it proves to be a big lump of work then don't bother - but in that case we should error if the orchestrator is anything other than "k8s"